### PR TITLE
support different SQL engines

### LINKED ISSUE #291: Migration SQL dialect
The `CreateMigrationsTable` class is written for SQLite, which isn't compatible with MySQL. Maybe we should provide three different implementations (SQLite, MySQL and postgres), the developer will need to publish the right migration file when setting up the database.

### LINKED ISSUE #297: [bug] sorting migrations are not done by name
**Problem**
It seems that the order in which migrations are run, are either not working as expected or I haven't got migration naming correct.

**Scenario**
When running two migrations, where the latter creates a constraint on the former.
The one with the constraint is run before the referenced table is created.

Running the two migrations separate is a work around in this case.

**Steps to reproduce**
- Create two migrations with the following statements
- run `php tempest migrate:up`
- the following error occurs
    `General error: 1824 Failed to open the referenced table 'authors`

<details>
<summary>Tabel 1 (authors)</summary>

```sql
# migration name: 1970-01-02_create_authors_table
create table if not exists authors (
    id bigint auto_increment primary key,
    name varchar(255) not null
);
```

</details>

<details>
<summary>Table 2 (books)</summary>

```sql
# migration name: 1970-01-03_create_books_table
create table books
(
    id        bigint auto_increment primary key,
    title     varchar(255) not null,
    author_id bigint       null,
    constraint books_authors_id_fk foreign key (author_id) references authors (id)
);

create index books__authors__index on books (author_id);
```

</details>

### LINKED ISSUE #298: [feature] add support for multiple database engines
resolves #179 #291

fixes #297

**Situation**
At the moment the initial migration is SQLite only. In order to support multiple database engines we need a way to load migrations with their own SQL dialect.

**Proposed changes**
Combine the above mentioned issues.

- [ ] create SQL files with the migration table creation, for every database driver we currently support
- [ ] the name of the migration should dictate the order that they are run in 
        (#297)
- [ ] let the `CreateMigrationTable` class decide which file to use, based on the currently configured database driver
- [ ] let user generated migrations use the same approach

### PR DESCRIPTION
**Pull request by Treggats**
resolves #298 
resolves #291
fixes #297

This pull request implements support for the supported database engines.
- MySQL
- PostgreSQL
- SQLite

It does so by having a "builder" class which only implements syntax that is different between the different engines. For example the primary key.

It also addresses the sorting of migrations. This caused a bug when dealing with migrations that have foreign keys defined.

**Todo:**
- [x] move binding the database driver to an initializer for the DatabaseDriver
- [x] move the `QueryStatement` init to the driver (`$this->driver->createQueryStatement()`)
- [x] improve the coverage for the `QueryStatement` class
- [x] move the `#[Singleton]` attribute to the `initialize` method (database driver initalizer)
- [x] move sorting the migrations to the manager
- [x] (optional) remove the need of the closure in the `QueryStatement` class
- [ ] (optional) make the database dialect configurable

### CODE REVIEW

File: src/Tempest/Database/Migrations/MigrationManager.php
Comment (Treggats) [inline src/Tempest/Database/Migrations/MigrationManager.php:30]: at the moment passing the driver to the migration, as I mentioned above I'm looking if I can skip this

File: src/Tempest/Database/QueryStatement.php
Comment (Treggats) [inline src/Tempest/Database/QueryStatement.php:76]: this will throw a match exception if the driver is not in this list. Which is what we want because that driver might have a different syntax and if so it needs to be added to this list.

File: src/Tempest/Database/QueryStatement.php
Comment (Treggats) [inline src/Tempest/Database/QueryStatement.php:98]: it would be nice if we could somehow validate the syntax per driver. Haven't looked into it yet though

File: src/Tempest/Framework/Application/Kernel.php
Comment (Treggats) [inline src/Tempest/Framework/Application/Kernel.php:40]: binding the current driver to the driver interface/contract. Currently figuring out if I can make this work for the `QueryStatement` class

File: src/Tempest/Framework/Testing/IntegrationTest.php
Comment (Treggats) [inline src/Tempest/Framework/Testing/IntegrationTest.php:54]: same as above, but for tests

File: tests/Unit/Database/DatabaseQueryStatementTest.php
Comment (Treggats) [inline tests/Unit/Database/DatabaseQueryStatementTest.php:15]: test to check the syntax manually

File: src/Tempest/Database/MigrationDiscovery.php
Comment (brendt) [reply src/Tempest/Database/MigrationDiscovery.php:41]: TIL!

File: src/Tempest/Database/Migrations/CreateMigrationsTable.php
Comment (brendt) [reply src/Tempest/Database/Migrations/CreateMigrationsTable.php:14]: You could add something like `$this->driver->createQueryStatement()`, so that you don't have to manually pass it in L26. I wouldn't bother with trying to be more clever than that for now.

File: src/Tempest/Framework/Application/Kernel.php
Comment (brendt) [reply src/Tempest/Framework/Application/Kernel.php:40]: You should move this logic outside the kernel and create an initializer class for `DatabaseDriver` 👍

File: src/Tempest/Framework/Testing/IntegrationTest.php
Comment (brendt) [reply src/Tempest/Framework/Testing/IntegrationTest.php:54]: This is ok to hard-code for tests

**Review by brendt — CHANGES_REQUESTED** [review]

File: src/Tempest/Database/DatabaseConfig.php
Comment (Treggats) [inline src/Tempest/Database/DatabaseConfig.php:32]: helper to get and sort the migrations.

File: src/Tempest/Database/DatabaseDriverInitializer.php
Comment (Treggats) [inline src/Tempest/Database/DatabaseDriverInitializer.php:12]: added wiring for the database driver, so I don't have to register it manually

File: src/Tempest/Database/Drivers/MySqlDriver.php
Comment (Treggats) [inline src/Tempest/Database/Drivers/MySqlDriver.php:42]: create the `QueryStatement` class through the driver

File: src/Tempest/Database/Migrations/MigrationManager.php
Comment (Treggats) [inline src/Tempest/Database/Migrations/MigrationManager.php:40]: get the migrations from start to end. e.g. 0 -> 100

File: src/Tempest/Database/Migrations/MigrationManager.php
Comment (Treggats) [inline src/Tempest/Database/Migrations/MigrationManager.php:68]: when reversing the table creation, we need the migration but in reverse.

Otherwise, if you use foreign keys, you'd get an constraint exception. Because the referenced table got dropped first, but has a reference elsewhere. 💥

File: src/Tempest/Database/DatabaseConfig.php
Comment (Treggats) [inline src/Tempest/Database/DatabaseConfig.php:34]: fixes #297

File: src/Tempest/Database/DatabaseConfig.php
Comment (brendt) [inline src/Tempest/Database/DatabaseConfig.php:10]: We make config properties public. 

In fact, we're actually gonna wait for PHP 8.4 until we release a v1 of Tempest, and we're going to refactor all interfaces to use property hooks instead of methods. It's not related to these properties, but I figured I'd let you know anyway :)

File: src/Tempest/Database/DatabaseConfig.php
Comment (brendt) [reply src/Tempest/Database/DatabaseConfig.php:32]: I guess you need to sort them both ways for up and down? I don't really like hard-coded strings. But it also feels weird to make an enum for it. A boolean also feels icky… Would it make sense to do the sorting in place (where the get method is called)? I reckon it's only one place, but I could be wrong…

File: src/Tempest/Database/DatabaseDriverInitializer.php
Comment (brendt) [reply src/Tempest/Database/DatabaseDriverInitializer.php:12]: FYI, you'll need to move the `#[Singleton]` attribute to the `initialize` method instead of the class. This was a change I made this morning :)

File: src/Tempest/Database/Drivers/MySqlDriver.php
Comment (brendt) [reply src/Tempest/Database/Drivers/MySqlDriver.php:42]: 👍

File: src/Tempest/Database/DatabaseConfig.php
Comment (brendt) [reply src/Tempest/Database/DatabaseConfig.php:32]: Also the naming is a bit weird, shouldn't it be `getMigrations`?

File: src/Tempest/Database/Migrations/MigrationManager.php
Comment (brendt) [reply src/Tempest/Database/Migrations/MigrationManager.php:68]: Circling back to my previous comment, I think I'd prefer to move the array sorting within the migration manager, so that we can get rid of that `get` method with its hard-coded string for asc/desc entirely.

File: src/Tempest/Database/QueryStatement.php
Comment (brendt) [inline src/Tempest/Database/QueryStatement.php:16]: FYI, slightly different code style here, but CS Fixer will take care of it.

File: src/Tempest/Database/QueryStatement.php
Comment (brendt) [inline src/Tempest/Database/QueryStatement.php:52]: What does `@coverage` do?

File: src/Tempest/Database/QueryStatement.php
Comment (brendt) [inline src/Tempest/Database/QueryStatement.php:46]: I think we might refactor at a later point to have dedicated grammar classes, though that's not necessary for now.

File: tests/Fixtures/Migrations/CreateAuthorTable.php
Comment (brendt) [inline tests/Fixtures/Migrations/CreateAuthorTable.php:26]: I wonder if we need the closure? 

```php
$this->driver
    ->createQueryStatement('Author')
    ->primary()
    ->statement('name TEXT NOT NULL')
    ->statement('type TEXT')
    ->toQuery();
```

**Review by brendt — CHANGES_REQUESTED** [review]

File: src/Tempest/Database/DatabaseConfig.php
Comment (Treggats) [reply src/Tempest/Database/DatabaseConfig.php:10]: Sounds interesting 😋 

for now I'll make it public.
```suggestion
        public readonly DatabaseDriver $driver,
```

File: src/Tempest/Database/DatabaseConfig.php
Comment (Treggats) [reply src/Tempest/Database/DatabaseConfig.php:32]: >  I don't really like hard-coded strings. But it also feels weird to make an enum for it. A boolean also feels icky

I feel the same way 😅 this in combination with the match expression seemed like a good idea. But perhaps I'll move the sorting the the migration manager.

File: src/Tempest/Database/QueryStatement.php
Comment (Treggats) [reply src/Tempest/Database/QueryStatement.php:52]: it was my part of remembering which part of the class needed tests :) forgot to remove it.
```suggestion
        $this->query[] = sprintf(
```

File: tests/Fixtures/Migrations/CreateAuthorTable.php
Comment (Treggats) [reply tests/Fixtures/Migrations/CreateAuthorTable.php:26]: the idea behind the closure is so that I can format the query easily. Maybe I can remove the need for it, but would require some work.

File: src/Tempest/Database/MigrationDiscovery.php
Comment (Treggats) [reply src/Tempest/Database/MigrationDiscovery.php:41]: I have a branch open where I do this for 99% of these cases

File: src/Tempest/Database/Migrations/MigrationManager.php
Comment (Treggats) [inline src/Tempest/Database/Migrations/MigrationManager.php:60]: this is the code that the table doesn't exist. Writing this I thought about getting the table name from the exception itself 🤔

File: src/Tempest/Database/Migrations/MigrationManager.php
Comment (Treggats) [inline src/Tempest/Database/Migrations/MigrationManager.php:73]: fixes #297

File: src/Tempest/Database/QueryStatement.php
Comment (Treggats) [reply src/Tempest/Database/QueryStatement.php:46]: I like that idea

File: src/Tempest/Console/Commands/DiscoveryClearCommand.php
Comment (brendt) [inline src/Tempest/Console/Commands/DiscoveryClearCommand.php:35]: This is actually needed during testing, however, because we call it during testing bootstrap, it only triggers once per test run

File: src/Tempest/Database/DatabaseConfig.php
Comment (brendt) [inline src/Tempest/Database/DatabaseConfig.php:11]: I guess this can be public again now that the getter and setter don't do anything? That means the getter and setter can also be removed

File: src/Tempest/Database/DatabaseDriver.php
Comment (brendt) [inline src/Tempest/Database/DatabaseDriver.php:17]: Maybe this can be moved to the Dialect enum?

File: src/Tempest/Database/DatabaseDriver.php
Comment (brendt) [inline src/Tempest/Database/DatabaseDriver.php:15]: Perhaps we can even move this to the `DatabaseDialect` enum, and simply have an `getDialect` on the driver class?

File: src/Tempest/Framework/Application/AppConfig.php
Comment (brendt) [inline src/Tempest/Framework/Application/AppConfig.php:17]: I think it'll be better to move this to the database driver class, as I mentioned in a previous comment

File: src/Tempest/Framework/Config/database.php
Comment (brendt) [inline src/Tempest/Framework/Config/database.php:16]: I don't think we need to provide this level of configurability: let's keep sqlite as the only option, if people want another driver, they should simply provide their own database config

File: tests/Fixtures/Migrations/CreateAuthorTable.php
Comment (brendt) [reply tests/Fixtures/Migrations/CreateAuthorTable.php:26]: Let's keep it for now, I'll probably look into refactoring it after the PR is merged :)

**Review by brendt — CHANGES_REQUESTED** [review]

File: src/Tempest/Database/DatabaseDriver.php
Comment (Treggats) [reply src/Tempest/Database/DatabaseDriver.php]: The `createQueryStatement(..)` needs a instance of the database driver. I'm not sure how to pass that through the enum.

**Comment by brendt** [comment]
💪 I think I'm good to merge this! Anything else you wanted to tackle before I do?

**Comment by Treggats** [comment]
> 💪 I think I'm good to merge this! Anything else you wanted to tackle before I do?

Nope, think I'm good. :)

File: src/Tempest/Database/Migrations/MigrationManager.php
Comment (Treggats) [inline src/Tempest/Database/Migrations/MigrationManager.php:41]: normal sort, would otherwise cause errors when dealing with foreign keys #297 
```suggestion
        $migrations = $this->databaseConfig->getMigrations();
        ksort($migrations);
        foreach ($migrations as $migrationClassName) {
```

**Comment by Treggats** [comment]
working on a additional test for the output given on the console. But that doesn't block this pull request.

So it's ready to merge 🎉

**Comment by brendt** [comment]
Let's merge!

**Pull request merged**
